server/sonar-server/src/main/java/org/sonar/server/permission/index/PermissionIndexer.java

```diff
@@ -132,48 +140,42 @@ public void indexProject(String projectUuid, Cause cause) {
   @Override
   public void deleteProject(String projectUuid) {
     authorizationScopes.forEach(scope -> esClient
-      .prepareDelete(scope.getIndexName(), TYPE_AUTHORIZATION, projectUuid)
+      .prepareDelete(scope.getIndexType(), projectUuid)
       .setRouting(projectUuid)
       .setRefresh(true)
       .get());
   }
 
-  private boolean isAuthorizationTypeEmpty(String index) {
-    SearchResponse response = esClient.prepareSearch(index).setTypes(TYPE_AUTHORIZATION).setSize(0).get();
-    return response.getHits().getTotalHits() == 0;
-  }
-
-  private void truncateAuthorizationType(String index) {
-    BulkIndexer.delete(esClient, index, esClient.prepareSearch(index).setTypes(TYPE_AUTHORIZATION).setQuery(matchAllQuery()));
+  private Stream<AuthorizationScope> getScopes(Set<IndexType> indexTypes) {
+    return authorizationScopes.stream()
+      .filter(scope -> indexTypes.contains(scope.getIndexType()));
   }
 
-  @VisibleForTesting
-  void index(Collection<PermissionIndexerDao.Dto> authorizations) {
+  private void index(Collection<PermissionIndexerDao.Dto> authorizations, Stream<AuthorizationScope> scopes) {
     if (authorizations.isEmpty()) {
       return;
     }
-    int count = 0;
-    BulkRequestBuilder bulkRequest = esClient.prepareBulk().setRefresh(false);
-    for (PermissionIndexerDao.Dto dto : authorizations) {
-      for (AuthorizationScope scope : authorizationScopes) {
-        if (scope.getProjectPredicate().test(dto)) {
-          bulkRequest.add(newIndexRequest(dto, scope.getIndexName()));
-          count++;
-        }
-      }
-      if (count >= MAX_BATCH_SIZE) {
-        EsUtils.executeBulkRequest(bulkRequest, BULK_ERROR_MESSAGE);
-        bulkRequest = esClient.prepareBulk().setRefresh(false);
-        count = 0;
-      }
-    }
-    if (count > 0) {
-      EsUtils.executeBulkRequest(bulkRequest, BULK_ERROR_MESSAGE);
-    }
-    authorizationScopes.forEach(type -> esClient.prepareRefresh(type.getIndexName()).get());
+
+    // index each authorization in each scope
+    scopes.forEach(scope -> index(authorizations, scope));
+  }
+
+  private void index(Collection<PermissionIndexerDao.Dto> authorizations, AuthorizationScope scope) {
+    IndexType indexType = scope.getIndexType();
+
+    BulkIndexer bulkIndexer = new BulkIndexer(esClient, indexType.getIndex());
+    bulkIndexer.setLarge(/* TODO */false);
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)